### ISSUE TITLE
Add kernels faster

### ISSUE THREAD

**Issue by YaqiWang** [idaholab/moose#2258]
When the number of kernels is large, possibly reaches 1M, the current way of adding kernels is not efficient. We should improve it. Also I think it is time to investigate what are stored in Kernel and try to make it perfect in term of memory efficiency. Other issues can be related with this ticket with large number of variables and kernels.

Thanks.

**Comment by YaqiWang** [comment]
Hopefully this has been fixed. I don't have that crazy test anymore which adds millions of kernels and calling this function millions of times.

**Comment by permcody** [comment]
Have you tried any timings yet? I'd be interested in seeing them, we are going to be doing more of these kinds of things.

**Comment by YaqiWang** [comment]
We can create an dummy action and add a million of kernels which is defined on a list of blocks based on a material property like diffusion coefficient. We will also know  how much memory this million of kernels will take. You prefer me to do this?

**Comment by YaqiWang** [comment]
Actually, there is one action under test called `AddLotsOfDiffusion`, we can change it to add a material property and add diffusion kernels using that property. If the time-out time is 300s, we make it run with 250s with the maximum number. This way, whenever there is a bad implementation slowing down the performance, we can be notified.

**Comment by permcody** [comment]
I'd rather make it run in debug in around 10-15s then set a timeout of 20s
or so for testing, unless of course it really needs to be that big but
there's no need to make it run for 5 mins.

**Comment by YaqiWang** [comment]
We can set a timeout for each individual test?

**Comment by permcody** [comment]
Yes, run ./run_tests --dump you'll see all available options for each
Tester type. There is one called "max_time" available for each type.

Cody

**Comment by YaqiWang** [comment]
@permcody Because the change is simple, I just paste my changes here:
```
diff --git a/test/src/actions/AddLotsOfDiffusion.C b/test/src/actions/AddLotsOfDiffusion.C
index 392e7ef..b74920e 100644
--- a/test/src/actions/AddLotsOfDiffusion.C
+++ b/test/src/actions/AddLotsOfDiffusion.C
@@ -41,6 +41,7 @@ InputParameters validParams<AddLotsOfDiffusion>()
 
   InputParameters params = validParams<AddVariableAction>();
   params.addRequiredParam<unsigned int>("number", "The number of variables to add");
+  params.addParam<unsigned int>("number_kernel", 1, "The number of kernels per variable to add");
 
   return params;
 }
@@ -55,6 +56,7 @@ void
 AddLotsOfDiffusion::act()
 {
   unsigned int number = getParam<unsigned int>("number");
+  unsigned int number_kernel = getParam<unsigned int>("number_kernel");
 
   if (_current_task == "add_variable")
   {
@@ -64,15 +66,31 @@ AddLotsOfDiffusion::act()
       addVariable(var_name);
     }
   }
+  else if (_current_task == "add_material")
+  {
+    InputParameters params = _factory.getValidParams("GenericConstantMaterial");
+    params.set<std::vector<std::string> >("prop_names").assign(1, "dc");
+    params.set<std::vector<Real> >("prop_values").assign(1, 1.0);
+    _problem->addMaterial("GenericConstantMaterial", "diffusion_coef", params);
+  }
   else if (_current_task == "add_kernel")
   {
     for (unsigned int cur_num = 0; cur_num < number; cur_num++)
     {
       std::string var_name = getShortName() + Moose::stringify(cur_num);
 
-      InputParameters params = _factory.getValidParams("Diffusion");
+      InputParameters params = _factory.getValidParams("DiffMKernel");
       params.set<NonlinearVariableName>("variable") = var_name;
-      _problem->addKernel("Diffusion", var_name, params);
+      for (unsigned int i = 0; i < number_kernel; i++)
+      {
+        std::string ker_name = var_name + "_" + Moose::stringify(i);
+
+        MaterialPropertyName dc = "dc";
+        params.set<std::vector<SubdomainName> >("block") = _problem->getMaterialPropertyBlockNames(dc);
+        params.set<MaterialPropertyName>("mat_prop") = dc;
+        _problem->addKernel("Diffusion", ker_name, params);
+      }
+      Moose::out << " " << number_kernel << " kernels for " << var_name << " are added" << std::endl;
     }
   }
   else if (_current_task == "add_bc")
```

I tried different numbers of kernels per variable with the existing test and commenting out the call of `getMaterialPropertyBlockNames()`:
```
n  time(s) 
1 6.59297 
2 7.04755
4 8.01655
8 10.3755
16 16.0182 
32 27.4201
```
uncomment the call with 32 kernel per var, I got 27.0969. Looks ok to me now.

**Comment by YaqiWang** [comment]
Actually that call does not test much after I looked into the code. We need to create a mesh with lots of blocks and then do the test again. We will probably see something much different. The algorithm seems still flawed.

**Comment by aeslaughter** [comment]
The memory issue that may or may not exist with InputParameters will not be fixed until I get #5564 merged, which will probably be next week when I return since there is integration issues.

**Comment by friedmud** [comment]
I think I'm closing this as "Won't fix"... I don't think there is anything we can do about it currently.

A better plan is just to add less Kernels (i.e. implement VectorKernels...).
